Allow users to get error locations

cetra3 opened the pull request:
Simple change to make ErrorImpl public.

I'd like to be able to use the Marker Information from the scanner error to provide users that are changing files the ability to check to see whether their yaml is valid and, if not, where the issue lies.  To do this i need the Marker information from the ScannerError in a structured format, so I can feed that into an editor.

This resolves https://github.com/dtolnay/serde-yaml/issues/78

I also have a [PR out](https://github.com/chyh1990/yaml-rust/pull/90) about getting access to the Marker for Scanner errors, which is also not public.

cetra3 commented:
Ok, I've added two new structs to the Error which decouple the Error from yaml_rust in the future.  They are `ErrorMarker` which is just `Marker`, and `ErrorPos` which is a description with an optional ErrorMarker. 

The ErrorPos struct is basically what I'm after, so I can get a nice location of where the error is happening (if one exists).

Unfortunately though, yaml_rust does not currently expose a way of getting the marker from a scan error, which I have submitted a PR for, but may not get accepted, so I've just pointed this PR to my branch.

But to make matters worse, the new version of yaml_rust escapes `y` as a key with quotes around (i.e, `"y"` instead of `y`) I think to align to the [bool](http://yaml.org/type/bool.html) definition within yaml which breaks assertions and examples.

dtolnay commented:
Please rebase on the master branch which includes the update to yaml-rust 0.4.0.

dtolnay commented on src/error.rs:
```diff
@@ -19,11 +19,63 @@ use yaml_rust::scanner::{self, Marker, ScanError};
+    }
+}
+
+impl<'a> From<&'a Marker> for ErrorMarker {
```
This is still yaml-rust in the public API of serde-yaml. Imagine this code:

```toml
[dependencies]
yaml-rust = "0.3"
serde_yaml = "0.7"
```

```rust
use yaml_rust::scanner::Marker;
use serde_yaml::ErrorMarker;

fn x(marker: &Marker) -> ErrorMarker {
    From::from(marker)
}
```

This code would fail to compile after we bump yaml-rust version.

cetra3 commented:
I've adjusted the from to be a private fn of ErrorMarker rather than implementing the From trait.  Is this enough to hide it?

cetra3 commented:
I've updated it to be a bit cleaner and provide some more documentation around usage.

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+        self.col
+    }
+
+    fn from_marker(marker: &Marker) -> Self {
```
Can you add a comment that this is required to keep Marker private?

Alternatively, you can impl From and add `#[doc(hidden)]`.

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// assert_eq!(marker.line(), 1);
+    /// assert_eq!(marker.col(), 0);
+    ///
+    ///
```
Why the empty lines?

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
+    /// Not all errors have a location, so this will return `None`
+    /// # Examples
```
Please add empty line above headline

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+  
 impl Error {
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
```
If you add an empty line above this one, the first line becomes a short and sweet summary in the API docs :)

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
 impl Error {
+    /// Returns the ErrorMarker from the error if one exists.
+    /// Allows for the exact location of the error to be gathered.
+    /// Not all errors have a location, so this will return `None`
```
Please end sentence with a period

killercup commented on src/error.rs:
````diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// ```rust
+    /// # extern crate serde_yaml;
+    /// # use serde_yaml::{Value, Error};
+    ///
````
As the lines above and below are hidden in the docs, this code block will start with an empty line. Speaking of empty lines, can you remove this one? 😉

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// # fn main() {
+    /// # 
+    /// // The `@` character as the first character makes this invalid yaml
+    /// let result: Result<Value, Error> = yaml(r#"@invalid_yaml"#);
```
No need to use a `r#"raw"#` string here, right?

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// # extern crate serde_yaml;
+    /// # use serde_yaml::{Value, Error};
+    ///
+    /// # fn yaml(i: &str) -> Result<Value, Error> { serde_yaml::from_str(i) }
```
I'm not sure this is necessary, as `yaml` is only ever used once in this example. And `serde_yaml::from_str` is not that verbose, is it?

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+        match *self.0 {
+            ErrorImpl::Message(_, Some(ref pos)) => Some(ErrorMarker::from_marker(&pos.marker)),
+            ErrorImpl::Scan(ref scan) => Some(ErrorMarker::from_marker(scan.marker())),
+            _ => None
```
pls add trailing comma

killercup commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    /// let marker = result.unwrap_err().marker().unwrap();
+    ///
+    /// assert_eq!(marker.line(), 1);
+    /// assert_eq!(marker.col(), 0);
```
It's weird to see lines start at 1 but cols at zero… but whatever

dtolnay commented:
How come the column does not match what you see when you print the error?

```rust
extern crate serde_yaml;
use serde_yaml::Value;

fn main() {
    let err = serde_yaml::from_str::<Value>("@").unwrap_err();
    let marker = err.marker().unwrap();
    println!("{}", err);
    println!("line={} col={}", marker.line(), marker.col());
}
```

```
unexpected character: `@' at line 1 column 1
line=1 col=0
```

dtolnay commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+
+/// This type represents the location that an error occured
+#[derive(Debug)]
+pub struct ErrorMarker {
```
Could you brainstorm some alternative names for this? [`marker`](https://doc.rust-lang.org/std/marker/index.html) means something else in Rust.

dtolnay commented on src/error.rs:
```diff
@@ -49,7 +49,72 @@ pub struct Pos {
+    }
+
+    /// The column of the error
+    pub fn col(&self) -> usize {
```
I would prefer not to abbreviate "column" because the standard library does not abbreviate it [here](https://doc.rust-lang.org/std/macro.column.html) and [here](https://doc.rust-lang.org/std/panic/struct.Location.html#method.column).

cetra3 commented:
@dtolnay I'm not sure about why the marker is different from the error. I'll have a look at the yaml crate and see if I can pin that one down.

@killercup I'll make those adjustments and check that it formats it as you said.

cetra3 commented:
I've pushed a new commit to the branch with some of those suggestions.

I've changed the name simply to `Location` and used `column()` rather than `col`.

With regards to the column not matching the error message, I've just added + 1 to what we get from the marker, as that is what they're doing when they print it out in the [yaml crate](https://github.com/chyh1990/yaml-rust/blob/master/src/scanner.rs#L83).

Also added a test case for the None side of things.

dtolnay commented:
I will get a release out tomorrow with this change and the update to yaml-rust 0.4.

cetra3 commented:
Awesome!
